### ISSUE TITLE
Flag duplicate constraints

### ISSUE THREAD

**Issue by d-e-s-o** [rust-lang/rust-clippy#8176]
### What it does

It would be nice to have a lint that flags duplicate constraints on generic parameters. For example,

```
struct Foo<T>(T);

impl<T> Foo<T>
where
  T: Ord,
{
  fn bar(&self)
  where
    T: Ord,
  {
  }
}
```
Here, `T` is constrained on both the `impl` and a method. There is nothing inherently wrong with it, but it's duplication and I can't think of a reason when it's desired. On the contrary, say the user wants to modify the constraint, now it may be easy to forget on location and end up with a big mess (which may or may not be detected early depending on whether constraints are loosened or not).

### Lint Name

duplicate-constraints

### Category

suspicious

### Advantage

- The lint would raise awareness that the user in all likelihood wants to specify a certain constraint only on the `impl` OR a method
- As alluded to above, having duplicate constraints may become a correctness issue in the presence of changes, if one location is forgotten to be adjusted
- It's also just less code and we get back to having a single source of truth

### Drawbacks

None known.

### Example

```rust
struct Foo<T>(T);

impl<T> Foo<T>
where
  T: Ord,
{
  fn bar(&self)
  where
    T: Ord,
  {
  }
}
```

Could be written as:

```rust
struct Foo<T>(T);

impl<T> Foo<T>{
  fn bar(&self)
  where
    T: Ord,
  {
  }
}
```

**Comment by camsteffen** [comment]
Related: #873 (I think these could be the same lint)
